### DISCUSSION TITLE
How to return multiple objects from plugins file for use with Configuration API and Code Coverage

### DISCUSSION THREAD

**Discussion by vallme2003** [cypress-io/cypress#14375] [Plugins] [answered]
Hi, Currently my cypress tests are using this example for [multi environment configuration](https://docs.cypress.io/api/plugins/configuration-api.html#Switch-between-multiple-configuration-files)  which has a method to return from plugins file. so i also want to measure code coverage for my application, but when referring to [documentation](https://docs.cypress.io/guides/tooling/code-coverage.html#Install-the-plugin), it needs a different object to be returned. From what i read, we can return only single object from the plugins file. How do i achieve having both multi environment and code coverage?

Thanks!

**Comment by bahmutov** [comment, marked as answer]
At the end of the day, you need to:
1. let the plugins register their tasks and events, thus you might need to pass the `on` and `config` object to each plugin that requires
2. return the final modified `config` file.

In our case, you want to first load the appropriate environment config, then send it to the code coverage's registration function

```js
// promisified fs module
const fs = require('fs-extra')
const path = require('path')

function getConfigurationByFile (file) {
  const pathToConfigFile = path.resolve('..', 'config', `${file}.json`)

  return fs.readJson(pathToConfigFile)
}

// plugins file
module.exports = (on, config) => {
  // accept a configFile value or use development by default
  const file = config.env.configFile || 'development'
  const envConfig = getConfigurationByFile(file)
  // now register the code coverage plugin
  require('@cypress/code-coverage/task')(on, envConfig)
  // notice that the code coverage plugin modifies the config
  // passed as a parameter, thus we can simply return it
  return envConfig
}
```

**Reply by vallme2003** [reply]
Thank You so much for the response @bahmutov. I tried with this approach, but i seem to be getting an error with cypress 6.5.0 as :
**_TypeError: Cannot set property 'codeCoverageTasksRegistered' of undefined_** 

My plugins.js has additional plugins like these:

```
 module.exports = (on, config) => {
  
   const file = config.env.configFile || "development";
 
   const envConfig = getConfigurationByFile(file);

   require('@cypress/code-coverage/task')(on, envConfig);
 
   on("before:browser:launch", (browser, launchOptions) => {
     launchOptions.args.push("--another-arg");
     return launchOptions;
   });

   on("task", percyHealthCheck);
 
   on('before:browser:launch', (browser = {}, launchOptions) => {
     if (browser.family === 'chromium' && browser.name !== 'electron') {
       launchOptions.args.push('--disable-dev-shm-usage');
     }
 
     return launchOptions;
   });
 
   addMatchImageSnapshotPlugin(on, envConfig);
 
   return envConfig;
};
```
